### PR TITLE
#11824 Android: bundled loading page eliminates cold-start WebView error

### LINKED ISSUE #11824: Android: transient Chromium error page on cold start (ERR_CONNECTION_REFUSED / ERR_CONNECTION_CLOSED)
## Symptom

On Android cold start (especially first install or after clearing app data), users briefly see Chromium's built-in WebView error page:

> **Web page not available** — `https://localhost:8000/android` — `net::ERR_CONNECTION_CLOSED` (or `net::ERR_CONNECTION_REFUSED`)

After a few seconds it disappears and the app loads normally. Both error variants have been seen in the field.

## Root cause

Capacitor's `BridgeActivity.onCreate` auto-loads the configured server URL (`https://localhost:8000/`) **in parallel with** our in-process Rust server starting up. On cold start the Rust server typically takes ~500 ms to bind on a fast emulator, and is known to take several seconds on real low-end devices (DB migrations, certificate generation, processor init, etc).

When the WebView's auto-load fires before the listener is bound, the kernel returns RST → Chromium paints its native `ERR_CONNECTION_REFUSED` error page **inside the WebView, hidden behind the splash**. When the splash dismisses and our `handleOnStart` poll completes, we call `webView.loadUrl(localUrl + "/android")` to load the real app, but Chromium's error page is briefly visible during the swap.

The `ERR_CONNECTION_CLOSED` variant is the same race at a slightly later phase — TCP accepts into the kernel backlog before actix's accept loop drains it (~25 ms window measured on emulator), and the WebView's TLS handshake stalls. Same root cause, same symptom; just different timing relative to Rust startup.

Diagnostic measurements confirming all of the above are in the linked PR description.

## Reproduction

1. Build & install the debug APK on an Android emulator.
2. `adb shell pm clear org.openmsupply.client` (forces uninitialised cold-start path).
3. `adb shell monkey -p org.openmsupply.client -c android.intent.category.LAUNCHER 1`.
4. Watch for the Chromium error page during the splash → app transition.

To force the bug reliably even on a fast device, the readiness poll budget in `NativeApi.java` can be temporarily reduced from `retryCount = 5` to `retryCount = 1`.

## Affected versions

Believed to affect all Android builds prior to the fix. The race exists regardless of device, but is more visible on slow devices and on first install.

## Fix

See the linked PR — adds a bundled inline "Starting omSupply…" loading page as the WebView's initial content so the live server URL is never requested until our poll confirms the server is up.

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #11824

## Summary

- The Android WebView briefly shows Chromium's native error page on cold start because Capacitor's `BridgeActivity` auto-loads `https://localhost:8000/` before the in-process Rust server has bound the port. This PR replaces that initial URL with a bundled `assets/native/loading_page.html` so the live server URL is never requested until the readiness poll confirms the server is up.
- The custom `ErrorPage` HTML is also externalised to `assets/native/error_page.html` (same `loadUrl` pattern; the unhappy-face SVG no longer lives as a 250-line Java string literal).
- `AppState.isServerReady` is renamed to `isWebViewReady` — what it actually gates now is splash dismissal based on the WebView having its initial content, not Rust server state.
- Java readiness retry budget restored to 30 attempts (cheap; the LoadingPage is the visible state during the wait, so erring long is free).
- Adds permanent startup timing logs (Rust DB migrations / certs / bind+spawn; Java poll attempt + `WebView onReceivedError` hook) for future field diagnostics. These are cheap and only fire once per cold start.

## Diagnostic findings backing the fix

Measured on a Pixel-class emulator (M-chip Mac, debug build):

- **Rust cold-start → listener bound:** ~525 ms on fresh install ("Empty database detected"), ~590 ms on warm DB. Dominated by DB migrations + reports upsert + GraphQL schema build + cert generate on first run.
- **Bind → accept-loop draining first connection:** ~25 ms (measured with a TCP self-loopback probe spawned after `tokio::spawn(running_server)`, since removed). This window is a possible source of the field `ERR_CONNECTION_CLOSED` symptom — the kernel accepts the TCP connect into its backlog before actix's accept loop reads from the socket, so the WebView's TLS handshake stalls.
- **The visible error came from Capacitor's auto-load, not from any of our code.** `BridgeActivity.onCreate` calls `loadUrl(server.url)` synchronously, before `NativeApi.handleOnStart` ever fires. Repro logs showed `WebView onReceivedError url=https://localhost:8000/ errorCode=-6 description=net::ERR_CONNECTION_REFUSED` firing 100–200 ms **before** the Rust server even bound the port.

Layer-1 fix in this PR makes the WebView start on a bundled inline page, eliminating the race. Two follow-on improvements remain for a later PR:
- Layer 2: move `setWebViewReady(true)` from the immediate `webView.post(...)` to an `onPageFinished` for the `/android` URL — would guarantee zero flash even during the LoadingPage → app transition.
- Layer 3: tighten the `SSLHandshakeException` triage in the Java poll (only treat `CertPathValidatorException`-caused failures as ready) — defence-in-depth.

## What the user sees now

image

Cold start:
- 0 – ~200 ms: native splash (background `#f2f2f5`)
- ~200 ms – server-ready: brand-coloured CSS spinner + "Starting omSupply…" on the same `#f2f2f5` background
- At 5 s of waiting: "Show Log File" button fades in (slow-startup escape hatch)
- Server ready: `loadUrl("/android")` swaps the spinner for the React app
- Server times out: the existing custom `ErrorPage` (unhappy face + URL + Show Log button) loads from `assets/native/error_page.html`

No `https://localhost:8000/` is requested before the poll declares ready; the Chromium native error page should no longer appear.

## Files touched

**Java**
- `MainActivity.java` — `loadUrl(LoadingPage.URL)` immediately after `super.onCreate(...)`, register `LoadingPageInject`, post `setWebViewReady(true)` so splash dismisses on the next UI message.
- `NativeApi.java` — initial-load check uses `isWebViewReady() && connectedServer != null` (URL is now a `file://` URL on cold start); retry budget back to 30; poll-attempt logging + new `localUrl + "/android"` logging.
- `LoadingPage.java` (new) — `JavascriptInterface showLogs()` + `public static final String URL`.
- `ErrorPage.java` — HTML string removed; constructor now takes the failed URL, exposed via new `JavascriptInterface getFailedUrl()`; `showLogs()` unchanged.
- `AppState.java` — rename `isServerReady` → `isWebViewReady` (with comment clarifying semantics).
- `ExtendedWebViewClient.java` — `onPageStarted` skips Capacitor JS injection for `file:///android_asset/native/` URLs and adds `onReceivedError` / `onReceivedHttpError` logging.

**Bundled assets**
- `assets/native/loading_page.html` — spinner, headline, deferred-reveal log button.
- `assets/native/error_page.html` — unhappy-face SVG, URL display populated via `ErrorPageInject.getFailedUrl()`, log button.

**Rust**
- `server/server/src/lib.rs` — timing logs around DB migrations, cert load, bind, accept-loop spawn. No behavioural change.

## Test plan

- [ ] `yarn android:build:debug` + reinstall on emulator
- [ ] Fresh install cold start (`adb shell pm clear org.openmsupply.client`) — verify native splash → loading spinner → `/android` app, no Chromium error page at any point
- [ ] Warm restart (background → foreground) — verify no flash of loading page; app remains on `/android`
- [ ] Stuck-server simulation (temporarily add a sleep before `bind_rustls_0_23` longer than the poll budget) — verify ErrorPage with unhappy-face SVG renders correctly, "Show Log File" opens an AlertDialog with `remote_server.log` contents
- [ ] Slow cold-start simulation (sleep ~5 s before bind) — verify the "Show Log File" button fades in on the LoadingPage after 5 s, and tapping it shows logs
- [ ] Verify `adb logcat -v time | grep omSupply` shows the new poll-attempt + WebView-error logs; `adb shell run-as org.openmsupply.client cat files/remote_server.log` shows the new timing lines

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: client/packages/android/app/src/main/assets/native/error_page.html [hunk 1]
```diff
@@ -0,0 +1,296 @@
+<!DOCTYPE html>
```
Comment (jmbrunskill) [inline client/packages/android/app/src/main/assets/native/error_page.html:1]: This is the same page, but jsut moved it out of the Java String, possibly should be spearate issue...

**Comment by ria8651** [comment]
One initial thought, if we're going to show the logs on potentially every launch is there any reason not to mount the log folder to a user accessible dir?

**Comment by jmbrunskill** [comment]
> One initial thought, if we're going to show the logs on potentially every launch is there any reason not to mount the log folder to a user accessible dir?

It is a small window where they'll be accessible. Definitely attackable I guess. I like this being consistent with existing patterns. I agree maybe we can consider that other approach as well.

**Review by ria8651 — APPROVED** [review]
LGTM, did a quick test of log fetching and checked the loading page. The code looks fine but I'm not sure what else we need to test. Could be some weird edge case but I can't think what else to test.

File: client/packages/android/app/src/main/assets/native/loading_page.html [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+          #e83b30,
+          rgba(232, 59, 48, 0)
+        );
+        -webkit-mask: radial-gradient(
+          circle at center,
+          transparent 58%,
+          #000 60%
+        );
+        mask: radial-gradient(circle at center, transparent 58%, #000 60%);
+        animation: spin 1s linear infinite;
```
Comment (ria8651) [inline client/packages/android/app/src/main/assets/native/loading_page.html:46]: Oh one nit, the spinner goes the wrong way XD

**Comment by jmbrunskill** [comment]
> LGTM, did a quick test of log fetching and checked the loading page. The code looks fine but I'm not sure what else we need to test. Could be some weird edge case but I can't think what else to test.

I think it's ok, lets make sure QA get a chance to test this out before we release too, with a read Prod APK.
